### PR TITLE
[FLINK-12667][runtime] Add JobID to TaskExecutorGateway#releasePartitions

### PR DESCRIPTION
**Pull request by zentol**
## What is the purpose of the change

Adds `JobID` as an argument to `TaskExecutorGateway#releasePartitions` to simplify bookkeeping on the taskmanager side.

Additionally adds tests for calls to said method originating in `Execution`.

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java [hunk 1]
```diff
@@ -293,6 +299,81 @@ public void testSlotAllocationCancellationWhenExecutionCancelled() throws Except
...
+	/**
+	 * Tests that the partitions are released in case of a execution cancellation after the execution is already finished.
+	 */
+	@Test
+	public void testPartitionReleaseOnCancelAfterFinished() throws Exception {
+		testPartitionReleaseAfterFinished(Execution::cancel);
+	}
+
+	/**
+	 * Tests that the partitions are released in case of a execution suspension after the execution is already finished.
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java]: an execution

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java [hunk 2]
```diff
@@ -687,7 +687,7 @@ else if (current == RUNNING || current == DEPLOYING) {
 			}
 
 			else if (current == FINISHED || current == FAILED) {
-				// nothing to do any more. finished failed before it could be cancelled.
+				// nothing to do any more. finished/failed before it could be cancelled.
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java]: unrelated change

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 3]
```diff
@@ -641,7 +641,7 @@ private void stopTaskExecutorServices() throws Exception {
 	}
 
 	@Override
-	public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
+	public void releasePartitions(JobID jobId, Collection<ResultPartitionID> partitionIds) {
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: a bit weird that we will leave `jobId` here unused, is the future change that needs it going to be so big?

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java]: thats why it is in a separate hotfix commit

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 3 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: it's not gonna be big, but this part is simply already done.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java [hunk 4]
```diff
@@ -293,6 +299,81 @@ public void testSlotAllocationCancellationWhenExecutionCancelled() throws Except
...
+	public void testPartitionReleaseOnSuspendAfterFinished() throws Exception {
+		testPartitionReleaseAfterFinished(Execution::suspend);
+	}
+
+	private void testPartitionReleaseAfterFinished(Consumer<Execution> postFinishedExecutionAction) throws Exception {
+		final Tuple2<JobID, Collection<ResultPartitionID>> releasedPartitions = Tuple2.of(null, null);
+		final SimpleAckingTaskManagerGateway taskManagerGateway = new SimpleAckingTaskManagerGateway();
+		taskManagerGateway.setReleasePartitionsConsumer(releasedPartitions::setFields);
+
+		final SimpleSlot slot = new SimpleSlot(
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java]: Maybe we could provide an utility for creating `SimpleSlot` in this class, because it would be reused for many tests.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java [hunk 5]
```diff
@@ -293,6 +299,81 @@ public void testSlotAllocationCancellationWhenExecutionCancelled() throws Except
...
+		final Execution execution = executionVertex.getCurrentExecutionAttempt();
+
+		execution.allocateAndAssignSlotForExecution(
+			slotProvider,
+			false,
+			LocationPreferenceConstraint.ALL,
+			Collections.emptySet(),
+			TestingUtils.infiniteTime());
+
+		execution.deploy();
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java]: Maybe we could further provide an utility for creating `Execution` like below:
`
private CompletableFuture<Execution> createExecution(
		TaskManagerGateway taskManagerGateway,
		JobVertex... vertices) throws Exception {

		SimpleSlot slot = new SimpleSlot(
			new SingleSlotTestingSlotOwner(),
			new LocalTaskManagerLocation(),
			0,
			taskManagerGateway);

		ProgrammedSlotProvider slotProvider = new ProgrammedSlotProvider(1);
		slotProvider.addSlot(vertices[0].getID(), 0, CompletableFuture.completedFuture(slot));

		ExecutionGraph executionGraph = ExecutionGraphTestUtils.createSimpleTestGraph(
			new JobID(),
			slotProvider,
			new NoRestartStrategy(),
			vertices);

		executionGraph.start(TestingComponentMainThreadExecutorServiceAdapter.forMainThread());

		ExecutionJobVertex executionJobVertex = executionGraph.getJobVertex(vertices[0].getID());
		ExecutionVertex executionVertex = executionJobVertex.getTaskVertices()[0];
		Execution execution = executionVertex.getCurrentExecutionAttempt();

		CompletableFuture<Execution> allocationFuture = execution.allocateAndAssignSlotForExecution(
			slotProvider,
			false,
			LocationPreferenceConstraint.ALL,
			Collections.emptySet(),
			TestingUtils.infiniteTime());

		return allocationFuture;
	}
`
Then we could get `Execution` from future, and further get `ExecutionVertex` and `ExecutionGraph` from `Execution`. Then this helper could be reused for many existing tests.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java [hunk 6]
```diff
@@ -293,6 +299,81 @@ public void testSlotAllocationCancellationWhenExecutionCancelled() throws Except
...
+		execution.switchToRunning();
+
+		// simulate a case where a cancel/suspend call is too slow and the task is already finished
+		// in this case we have to explicitly release the finished partition
+		// if the task were canceled properly the TM would release the partition automatically
+		execution.markFinished();
+		postFinishedExecutionAction.accept(execution);
+
+		assertEquals(executionGraph.getJobID(), releasedPartitions.f0);
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java]: maybe use `assertThat` instead

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java [hunk 7]
```diff
@@ -293,6 +299,81 @@ public void testSlotAllocationCancellationWhenExecutionCancelled() throws Except
...
+		assertEquals(executionVertex.getProducedPartitions().size(), releasedPartitions.f1.size());
+		for (ResultPartitionID partitionId : releasedPartitions.f1) {
+			IntermediateResultPartition intermediateResultPartition = executionVertex
+				.getProducedPartitions()
+				.get(partitionId.getPartitionId());
+			assertNotNull(intermediateResultPartition);
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java]: this assert might be not necessary

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java [hunk 8]
```diff
@@ -293,6 +299,81 @@ public void testSlotAllocationCancellationWhenExecutionCancelled() throws Except
 		assertThat(canceledSlotRequests, equalTo(slotRequests));
 	}
 
+	/**
+	 * Tests that the partitions are released in case of an execution cancellation after the execution is already finished.
+	 */
+	@Test
+	public void testPartitionReleaseOnCancelAfterFinished() throws Exception {
```
Comment (gaoyunhaii) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java]: I think ...OnCancelingAfterFinished would be better in considering of the gramma, but I think I won't argue for that. :)

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java [hunk 9]
```diff
@@ -293,6 +299,81 @@ public void testSlotAllocationCancellationWhenExecutionCancelled() throws Except
...
+		testPartitionReleaseAfterFinished(Execution::cancel);
+	}
+
+	/**
+	 * Tests that the partitions are released in case of an execution suspension after the execution is already finished.
+	 */
+	@Test
+	public void testPartitionReleaseOnSuspendAfterFinished() throws Exception {
```
Comment (gaoyunhaii) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java]: Similarly, I think "...OnSuspendingAfterFinished" would be better, but I won't argue for that too.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java (hunk 4 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java]: such a utility already exists in #createProgrammedSlotProvider, will update the test to use that instead

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java (hunk 7 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java]: this check is done to ensure that the ids of all released partitions are actually valid. without it the test would pass even if completely random partitions were passed to the task executor.

I'll add a comment to clarify this; I had to think for a bit myself as to what we're checking here.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java (hunk 5 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionTest.java]: I agree that we could de-duplicate some code here. I'm concerned that we're baking in a few assumptions (vertex parallelism should be 1, first vertex in the array is special), and there a few subtle differences in tests that where I don't know whether they are significant or not. (For example, `#testTaskRestoreStateIsNulledAfterDeployment` doesn't allocate a slot beforehand). I don't have the time right now to really look into these things, so I'd move any larger refactoring to this class into a follow-up.
